File: hidden-api-stub/src/main/java/android/content/IContentProvider.java
package android.content;

import android.os.Bundle;
import android.os.IInterface;
import android.os.RemoteException;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

public interface IContentProvider extends IInterface {
    @RequiresApi(29)
    Bundle call(String callingPkg, String authority, String method,
                String arg, Bundle extras) throws RemoteException;

    @RequiresApi(30)
    Bundle call(String callingPkg, String attributionTag, String authority,
                String method, String arg, Bundle extras) throws RemoteException;

    @RequiresApi(31)
    Bundle call(AttributionSource attributionSource, String authority,
                String method, String arg, Bundle extras) throws RemoteException;

    Bundle call(String callingPkg, String method, @Nullable String arg, @Nullable Bundle extras)
            throws RemoteException;
}

File: server/src/main/java/me/gm/cleaner/server/ktx/IContentProvider.kt
package me.gm.cleaner.server.ktx

import android.content.AttributionSource
import android.content.IContentProvider
import android.os.Bundle
import android.os.RemoteException
import android.system.Os
import api.util.BuildUtils

@Throws(RemoteException::class)
fun IContentProvider.callCompat(
    callingPkg: String?, authority: String?, method: String?, arg: String?, extras: Bundle?
): Bundle {
    return when {
        BuildUtils.atLeast31() -> {
            try {
                call(
                    AttributionSource.Builder(Os.getuid()).setPackageName(callingPkg).build(),
                    authority, method, arg, extras
                )
            } catch (e: Throwable) {
                e.printStackTrace()
                call(callingPkg, null, authority, method, arg, extras)
            }
        }
        BuildUtils.atLeast30() -> {
            call(callingPkg, null, authority, method, arg, extras)
        }
        BuildUtils.atLeast29() -> {
            call(callingPkg, authority, method, arg, extras)
        }
        else -> {
            call(callingPkg, method, arg, extras)
        }
    }
}
